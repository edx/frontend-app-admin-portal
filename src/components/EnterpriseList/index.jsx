import React from 'react';
import PropTypes from 'prop-types';
import { Link, Redirect } from 'react-router-dom';
import Helmet from 'react-helmet';
import qs from 'query-string';

import H1 from '../H1';
import TableContainer from '../../containers/TableContainer';
import LmsApiService from '../../data/services/LmsApiService';
import SearchBar from '../SearchBar';

class EnterpriseList extends React.Component {
  constructor(props) {
    super(props);

    const { location } = props;
    const queryParams = qs.parse(location.search);
    this.state = {
      searchQuery: queryParams.search || '',
      searchSubmitted: !!queryParams.search,
    };
  }

  componentDidMount() {
    this.props.clearPortalConfiguration();
    this.props.getLocalUser();
  }

  handleSearch(query) {
    this.setState({
      searchQuery: query,
      searchSubmitted: true,
    });
    this.props.searchEnterpriseList({
      // Ternary used to turn empty string ('') into undefined so it does not get set in qs
      search: query ? query : undefined, // eslint-disable-line no-unneeded-ternary
    });
  }

  handleClear() {
    this.setState({
      searchQuery: '',
      searchSubmitted: true,
    });
  }

  formatEnterpriseData = enterprises => enterprises.map(enterprise => ({
    link: <Link to={`/${enterprise.slug}/admin`}>{enterprise.name}</Link>,
    name: enterprise.name,
    slug: enterprise.slug,
    uuid: enterprise.uuid,
  }));

  fetchEnterprisesWithSearch = (options) => {
    const optionsWithSearch = {
      ...options,
      // Ternary used to turn empty string ('') into undefined so it does not get set in qs
      search: this.state.searchQuery ? this.state.searchQuery : undefined,
    };

    return LmsApiService.fetchEnterpriseList(optionsWithSearch);
  }

  shouldRenderRedirectToEnterpriseAdminPage() {
    const {
      enterprisesData,
      loading,
      error,
    } = this.props;
    const { searchSubmitted } = this.state;
    const enterprises = enterprisesData && enterprisesData.results;

    return (
      !loading && !error && enterprises && enterprises.length === 1 &&
        enterprisesData.num_pages === 1 && !searchSubmitted
    );
  }

  renderRedirectToEnterpriseAdminPage() {
    const { results } = this.props.enterprisesData;
    return (
      <Redirect to={`/${results[0].slug}/admin`} />
    );
  }

  render() {
    const columns = [
      {
        label: 'Enterprise',
        key: 'link',
      },
    ];

    const { searchQuery } = this.state;

    return (
      <div>
        <Helmet>
          <title>Enterprise List</title>
        </Helmet>
        <div className="container">
          <div className="row mt-4">
            <div className="col-sm-12 col-md">
              <H1>Enterprise List</H1>
            </div>
            <div className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0">
              <SearchBar
                onSearch={query => this.handleSearch(query)}
                onClear={() => this.handleClear()}
                value={searchQuery}
              />
            </div>
          </div>
          <div className="row mt-2">
            <div className="col">
              {this.shouldRenderRedirectToEnterpriseAdminPage() &&
                this.renderRedirectToEnterpriseAdminPage()
              }
              <TableContainer
                id="enterprise-list"
                className="enterprise-list"
                fetchMethod={this.fetchEnterprisesWithSearch}
                columns={columns}
                formatData={this.formatEnterpriseData}
                tableSortable
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}


EnterpriseList.defaultProps = {
  enterprisesData: null,
  location: {
    search: null,
  },
  loading: false,
  error: null,
};

EnterpriseList.propTypes = {
  clearPortalConfiguration: PropTypes.func.isRequired,
  getLocalUser: PropTypes.func.isRequired,
  searchEnterpriseList: PropTypes.func.isRequired,
  enterprisesData: PropTypes.shape({
    count: PropTypes.number,
    num_pages: PropTypes.number,
    current_page: PropTypes.number,
    results: PropTypes.array,
    next: PropTypes.string,
    previous: PropTypes.string,
    start: PropTypes.number,
  }),
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
  loading: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
};

export default EnterpriseList;
