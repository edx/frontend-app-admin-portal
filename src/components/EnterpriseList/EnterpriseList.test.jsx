import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MemoryRouter, Redirect } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import EnterpriseList from './index';
import mockEnterpriseList from './EnterpriseList.mocks';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  paginateTable: () => {},
  sortTable: () => {},
  portalConfiguration: {
    enterpriseId: 'test-enterprise-id',
  },
  table: {
    enrollments: {
      data: {
        results: [],
        current_page: 1,
        num_pages: 1,
      },
      ordering: null,
      loading: false,
      error: null,
    },
  },
});

const EnterpriseListWrapper = props => (
  <MemoryRouter>
    <Provider store={store}>
      <EnterpriseList
        enterprisesData={{
          results: [],
        }}
        searchEnterpriseList={() => {}}
        clearPortalConfiguration={() => {}}
        getLocalUser={() => {}}
        {...props}
      />
    </Provider>
  </MemoryRouter>
);

describe('<EnterpriseList />', () => {
  let wrapper;

  describe('renders correctly', () => {
    it('calls getEnterpriseList, clearPortalConfiguration and getLocalUser props', () => {
      const mockClearPortalConfiguration = jest.fn();
      const mockGetLocalUser = jest.fn();
      const tree = renderer
        .create((
          <EnterpriseListWrapper
            clearPortalConfiguration={mockClearPortalConfiguration}
            getLocalUser={mockGetLocalUser}
          />
        ))
        .toJSON();
      expect(mockClearPortalConfiguration).toHaveBeenCalled();
      expect(mockGetLocalUser).toHaveBeenCalled();
      expect(tree).toMatchSnapshot();
    });

    it('with enterprises data', () => {
      const tree = renderer
        .create((
          <EnterpriseListWrapper
            enterprisesData={mockEnterpriseList}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with empty enterprises data', () => {
      const tree = renderer
        .create((
          <EnterpriseListWrapper
            enterprisesData={{
              ...mockEnterpriseList,
              count: 0,
              results: [],
            }}
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with search query and empty enterprises data', () => {
      const tree = renderer
        .create((
          <EnterpriseListWrapper
            enterprisesData={{
              ...mockEnterpriseList,
              count: 0,
              results: [],
            }}
            searchQuery="enterprise name"
            searchSubmitted
          />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with error state', () => {
      const tree = renderer
        .create((
          <EnterpriseListWrapper error={Error('Network Error')} />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with loading state', () => {
      const tree = renderer
        .create((
          <EnterpriseListWrapper loading />
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('redirects when there is only one enterprise', () => {
      const oneEnterpriseListData = {
        count: 1,
        current_page: 1,
        num_pages: 1,
        next: null,
        previous: null,
        results: [{
          uuid: 'ee5e6b3a-069a-4947-bb8d-d2dbc323396c',
          name: 'Enterprise 99',
          slug: 'enterprise-99',
        }],
        start: 0,
      };

      wrapper = mount((
        <MemoryRouter initialEntries={['/test']}>
          <Provider store={store}>
            <EnterpriseList
              enterprisesData={oneEnterpriseListData}
              searchEnterpriseList={() => {}}
              clearPortalConfiguration={() => {}}
              getLocalUser={() => {}}
            />
          </Provider>
        </MemoryRouter>
      ));
      const expectedRedirect = <Redirect to="/enterprise-99/admin/learners" />;
      expect(wrapper.containsMatchingElement(expectedRedirect)).toEqual(true);
    });
  });

  describe('enterprise list search', () => {
    const submitSearch = (searchQuery) => {
      wrapper.find('SearchBar').find('input[type=\'search\']').simulate('change', { target: { value: searchQuery } });
      expect(wrapper.find('SearchBar').find('input[type=\'search\']').prop('value')).toEqual(searchQuery);
      wrapper.find('SearchBar').find('.input-group-append').find('button').last()
        .simulate('click');
      expect(wrapper.find('EnterpriseList').instance().state.searchQuery).toEqual(searchQuery);
      expect(wrapper.find('EnterpriseList').instance().state.searchSubmitted).toBeTruthy();
    };

    it('searchQuery state changes onSearch', () => {
      wrapper = mount((
        <EnterpriseListWrapper
          enterprises={mockEnterpriseList}
        />
      ));

      submitSearch('Enterprise 1');
    });

    it('searchQuery state changes onClear', () => {
      wrapper = mount((
        <EnterpriseListWrapper
          enterprises={mockEnterpriseList}
        />
      ));

      submitSearch('Enterprise 1');

      wrapper.find('SearchBar').find('.input-group-append').find('button').first()
        .simulate('click');
      expect(wrapper.find('SearchBar').find('input[type=\'search\']').prop('value')).toEqual('');
      expect(wrapper.find('EnterpriseList').instance().state.searchQuery).toEqual('');
    });

    it('dispatches enterprise list request action when searchQuery prop changes', () => {
      const searchQuery = 'enterprise name';
      const mockGetEnterpriseList = jest.fn();
      wrapper = shallow((
        <EnterpriseListWrapper
          enterprises={mockEnterpriseList}
          getEnterpriseList={mockGetEnterpriseList}
        />
      )).find('EnterpriseList');

      wrapper.dive().setProps({
        searchQuery,
      }, () => {
        expect(wrapper.dive().state('searchQuery')).toEqual(searchQuery);
        expect(mockGetEnterpriseList).toHaveBeenCalledTimes(1);
      });
    });
  });
});
