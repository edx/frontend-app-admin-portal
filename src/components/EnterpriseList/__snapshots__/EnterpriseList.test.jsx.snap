// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EnterpriseList /> renders correctly calls getEnterpriseList, clearPortalConfiguration and getLocalUser props 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput2"
              id="label-asInput2"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput2"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput2"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput2"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EnterpriseList /> renders correctly with empty enterprises data 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput4"
              id="label-asInput4"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput4"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput4"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput4"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EnterpriseList /> renders correctly with enterprises data 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput3"
              id="label-asInput3"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput3"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput3"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput3"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EnterpriseList /> renders correctly with error state 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput6"
              id="label-asInput6"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput6"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput6"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput6"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EnterpriseList /> renders correctly with loading state 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput7"
              id="label-asInput7"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput7"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput7"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput7"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<EnterpriseList /> renders correctly with search query and empty enterprises data 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row mt-4"
    >
      <div
        className="col-sm-12 col-md"
      >
        <h1
          className={null}
        >
          Enterprise List
        </h1>
      </div>
      <div
        className="col-sm-12 col-md-6 col-lg-4 mb-3 mb-md-0"
      >
        <div
          className="border search-field"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <div
            className="form-group form-inline"
          >
            <label
              className=""
              htmlFor="asInput5"
              id="label-asInput5"
            >
              Search:
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              />
              <input
                aria-describedby="error-asInput5"
                aria-invalid={false}
                autoComplete="off"
                className="form-control is-invalid-nodanger input no-clear-btn"
                disabled={false}
                id="asInput5"
                name="search"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder=""
                readOnly={false}
                required={false}
                themes={Array []}
                type="search"
                value=""
              />
              <div
                className="input-group-append"
              >
                <button
                  className="btn search-btn"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  type="button"
                >
                  <span>
                    <span
                      aria-hidden={true}
                      className="fa fa-search"
                      id="Icon1"
                    />
                    <span
                      className="sr-only"
                    >
                      Submit search
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              aria-live="polite"
              className="invalid-feedback invalid-feedback-nodanger"
              id="error-asInput5"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
